Allow filtering params based on parent keys

gmalette opened the pull request:
Some parameters on APIs have ambiguous names
that cannot be filtered altogether.

For example, it makes sense not to filter :code in

```
{ file: { code: '<% source %>' }}
```

It makes sense however to filter it in

```
{ credit_card: { code: '424242424242' }}
```

This PR allows doing so by using 

```
config.filter_parameters += ["credit_card.code"]
```

gmalette committed:
Allow filtering params based on parent keys

Add the possibility to only filter parameters based on
their full path instead of relying on the immediate key.

    config.filter_parameters += ['credit_card.code']

    { 'credit_card' => { 'code' => '[FILTERED]' },
      'source' => { 'code' => '<%= puts 5 %>' } }

gmalette commented:
Has any thought been given to this? We've been running it for a while now

rafaelfranca commented:
cc @tenderlove @jeremy

arthurnn commented:
What is the slowest part of the new code? is this going to effect all apps, or only the ones that use this new feature?

gmalette commented:
> What is the slowest part of the new code?

The slowest part of the new code is the [line](https://github.com/rails/rails/pull/13897/files#diff-3db29ff2b03741b74d0c23eb59b7e19eR61) that assembles the new key and runs the regex. People not using the feature will not be impacted. The slowdown for people not using the feature comes from the [management of parent keys](https://github.com/rails/rails/pull/13897/files#diff-3db29ff2b03741b74d0c23eb59b7e19eR58) for each new level of nesting. As shown in the current benchmarks, the impact is minimal. 

Here are benchmarks for Ruby 2.2.1 (same code, but using IPS)

```
SMALL
Calculating -------------------------------------
     vanilla           844.000  i/100ms
     modified no .     725.000  i/100ms
     modified with .   658.000  i/100ms
-------------------------------------------------
     vanilla              8.838k (± 2.6%) i/s -     44.732k
     modified no .        8.705k (± 3.0%) i/s -     44.225k
     modified with .      7.016k (± 2.9%) i/s -     35.532k


MEDIUM
Calculating -------------------------------------
     vanilla             8.000  i/100ms
     modified no .       8.000  i/100ms
     modified with .     7.000  i/100ms
-------------------------------------------------
     vanilla             86.996  (± 3.4%) i/s -    440.000
     modified no .       85.721  (± 3.5%) i/s -    432.000
     modified with .     70.258  (± 4.3%) i/s -    357.000

LARGE
Calculating -------------------------------------
     vanilla             1.000  i/100ms
     modified no .       1.000  i/100ms
     modified with .     1.000  i/100ms
-------------------------------------------------
     vanilla              0.863  (± 0.0%) i/s -      5.000  in   5.795388s
     modified no .        0.849  (± 0.0%) i/s -      5.000  in   5.888606s
     modified with .      0.651  (± 0.0%) i/s -      4.000  in   6.169669s
```

arthurnn commented:
@gmalette Sounds good to me. Can you add a changelog before I merge it?
thanks

gmalette commented:
updated

arthurnn commented:
Merci beaucoup .
